src/main/java/net/fabricmc/loader/api/config/ConfigDefinition.java

```diff
@@ -0,0 +1,208 @@
+		this(namespace, name, version, serializer, saveType, upgrade, Paths.get(namespace, path), data);
+	}
+
+	/**
+	 * @param namespace namespace of the entity that owns this config file, usually a mod id
+	 * @param name the name of the config file this key represents (without any file extensions)
+	 * @param saveType see {@link SaveType}
+	 * @param version
+	 * @param upgrade
+	 */
+	public ConfigDefinition(@NotNull String namespace, @NotNull String name, @NotNull ConfigSerializer<R> serializer, @NotNull SaveType saveType, Map<DataType<?>, Collection<Object>> data, SemanticVersion version, ConfigUpgrade<R> upgrade) {
+		this(namespace, name, version, serializer, saveType, upgrade, Paths.get("."), data);
+	}
+
+	/**
+	 * @param namespace namespace of the entity that owns this config file, usually a mod id
+	 * @param saveType see {@link SaveType}
+	 * @param version
+	 * @param upgrade
+	 */
+	public ConfigDefinition(@NotNull String namespace, @NotNull ConfigSerializer<R> serializer, @NotNull SaveType saveType, Map<DataType<?>, Collection<Object>> data, SemanticVersion version, ConfigUpgrade<R> upgrade) {
+		this(namespace, "config", version, serializer, saveType, upgrade, Paths.get("."), data);
+	}
+
+	@SuppressWarnings("unchecked")
+	public <D> ListView<D> getData(DataType<D> dataType) {
+		return new ListView<>((List<D>) this.data.getOrDefault(dataType, Collections.emptyList()));
+	}
+
+	public @NotNull String getNamespace() {
+		return this.namespace;
+	}
+
+	public @NotNull String getName() {
+		return this.name;
+	}
+
+	public @NotNull ConfigSerializer<R> getSerializer() {
+		return this.serializer;
+	}
+
+	public @NotNull Path getPath() {
+		return this.path;
+	}
+
+	@Override
+	public boolean equals(Object o) {
+		if (this == o) return true;
+		if (o == null || getClass() != o.getClass()) return false;
+		ConfigDefinition<?> configDefinition = (ConfigDefinition<?>) o;
+		return namespace.equals(configDefinition.namespace) && name.equals(configDefinition.name);
+	}
+
+	@Override
+	public int hashCode() {
+		return Objects.hash(namespace, name);
+	}
+
+	@Override
+	public String toString() {
+		return this.string;
+	}
+
+	public static boolean isValid(String string) {
+		for(int i = 0; i < string.length(); ++i) {
+			if (!isCharacterValid(string.charAt(i))) {
+				return false;
+			}
+		}
+
+		return true;
+	}
+
+	private static boolean isCharacterValid(char c) {
+		return c == '_' || c == '-' || c >= 'a' && c <= 'z' || c >= '0' && c <= '9' || c == '.';
+	}
+
+	@Override
+	public int compareTo(@NotNull ConfigDefinition o) {
+		int c = this.namespace.compareTo(o.namespace);
+
+		if (c == 0) {
+			return this.name.compareTo(o.name);
+		} else {
+			return c;
+		}
+	}
+
+	public SaveType getSaveType() {
+		return this.saveType;
+	}
+
+	public boolean upgrade(@Nullable SemanticVersion from, R representation) {
+		return this.upgrade.upgrade(from, representation);
```

Comment:
Why do we need an upgrade api at all?

Reply:
If you change the config format (for example restructure to subobjects/subgroups), the old config file won't work properly.

Reply:
If we do have such a breaking change, should we even care about porting old configs? Many of them may simply be obsolete.

Reply:
Simple example, say you have
```java
public final class TechModConfig {
    public boolean enableOreGeneration;
    public boolean generateOreA, generateOreB, generateOreC, generateOreD;
}
```
because you only added the minimal config to allow modpack makers to configure your mod a little.
Then later you may want to move all the worldgen-related stuff to a subclass while keeping the old parameters.